elasticdl/python/worker/task_data_service.py

```diff
@@ -27,44 +32,77 @@ def __init__(
+            record_failure_tolerance_percentage
+        )
 
     def _reset(self):
         """
         Reset pending tasks and record counts
         """
-        self._record_count = 0
+
+        # Counters are keyed by task id.
+        self._cur_record_count = 0
         self._reported_record_count = 0
-        self._pending_tasks_with_counts = []
+        self._failed_record_count = 0
+        self._pending_tasks = deque()
         self._current_task = None
 
     def get_current_task(self):
         return self._current_task
 
+    def report_task_done(self, task_id, err_msg):
+        self._worker.report_task_result(task_id, err_msg)
+
+    def handle_next_task(self, err_msg=""):
+        with self._lock:
+            task = self._pending_tasks.popleft()
+            self._reported_record_count = 0
+            self._failed_record_count = 0
+            if self._pending_tasks:
+                self._current_task = self._pending_tasks[0]
+            self.report_task_done(task.task_id, err_msg)
+
     def report_record_done(self, count, err_msg=""):
         """
         Report the number of records in the latest processed batch,
         so TaskDataService knows if some pending tasks are finished
         and report_task_result to the master.
-        self._pending_tasks_with_counts[0][0] is the first pending task,
-        self._pending_tasks_with_counts[0][1] is the number of records
-        in this task.
         """
-        self._reported_record_count += count
-        if (
-            len(self._pending_tasks_with_counts)
-            and self._reported_record_count
-            >= self._pending_tasks_with_counts[0][1]
-        ):
-            with self._lock:
-                while (
-                    len(self._pending_tasks_with_counts)
-                    and self._reported_record_count
-                    >= self._pending_tasks_with_counts[0][1]
+        if self._pending_tasks:
+            task = self._pending_tasks[0]
+            self._reported_record_count += count
+            if err_msg:
+                self._failed_record_count += count
+
+            total_record_num = task.end - task.start
+            should_handle_next = False
+            if task.type == elasticdl_pb2.TRAINING:
+                # only training, we allow for failures
+                if (
+                    self._failed_record_count / total_record_num
+                    > self._record_failure_tolerance_percentage
                 ):
-                    task, _ = self._pending_tasks_with_counts.pop(0)
-                    self._worker.report_task_result(task.task_id, err_msg)
-                if len(self._pending_tasks_with_counts):
-                    self._current_task = self._pending_tasks_with_counts[0][0]
+                    should_handle_next = True
+            elif self._failed_record_count != 0:
+                should_handle_next = True
+            if self._reported_record_count >= total_record_num:
```

Comment:
`self._reported_record_count` is total records of all tasks, but `total_record_num` is only for the current pending tasks?
task A with 8 records reports finished 8 records: `self._reported_record_count = 8`, `total_record_num=8`.
Then task B with 8 records reports finished 4 records: `self._reported_record_count = 12`, `total_record_num= 8.
Task B is finished under this condition.

Reply:
The previous design is not correct. We can not use `self._reported_record_count`(current total report) > sum_of_all_record_num. Because we have to make sure that the records of a specific task are all correctly processed. Currently, there can be only one task being processed at the same time. So it is safe to directly use `self._reported_record_count` > `total_record_num` (which means the total num of the task being processed).

Reply:
In the previous design,  for pending tasks, if uses `while` loop to determine:
`self._reported_record_count`(current total report) > sum_of_all_record_num_till_this pending_task. 
So it can handle any situation for task size and batch size combination.

In current PR,  `self._reported_record_count` is reset to 0 after a task done, and `if` instead of `while` is used, so it may get wrong when:
1. `self._reported_record_count = 20` after adding the count, and pending_task's record number is 16. Then `self._reported_record_count` is reset to 0, but it should be 4 instead.
2. If task's record number is small and the minibatch number is larger, for example, each task has 32 records, minibatch=64, one minibatch will finish 2 tasks. So `while` loop should be used instead of `if`.